packages/fx-core/src/plugins/resource/bot/configs/teamsBotConfig.ts

```diff
@@ -23,60 +23,60 @@ export class TeamsBotConfig {
   public actRoles: PluginActRoles[] = [];
   public resourceNameSuffix = "";
 
-  public async restoreConfigFromContext(context: PluginContext): Promise<void> {
-    await this.scaffold.restoreConfigFromContext(context);
-    await this.provision.restoreConfigFromContext(context);
-    await this.localDebug.restoreConfigFromContext(context);
-    await this.deploy.restoreConfigFromContext(context);
+  private constructor() {
+    this.scaffold = new ScaffoldConfig();
+    this.provision = new ProvisionConfig();
+    this.localDebug = new LocalDebugConfig();
+    this.deploy = new DeployConfig();
+  }
+
+  public static async restoreConfigFromContext(context: PluginContext): Promise<TeamsBotConfig> {
+    let teamsBotConfig = new TeamsBotConfig();
+    await teamsBotConfig.scaffold.restoreConfigFromContext(context);
+    await teamsBotConfig.provision.restoreConfigFromContext(context);
+    await teamsBotConfig.localDebug.restoreConfigFromContext(context);
+    await teamsBotConfig.deploy.restoreConfigFromContext(context);
 
-    const clientIdValue: ConfigValue = context.configOfOtherPlugins
+    teamsBotConfig.teamsAppClientId = context.configOfOtherPlugins
       .get(PluginAAD.PLUGIN_NAME)
       ?.get(PluginAAD.CLIENT_ID);
-    if (clientIdValue) {
-      this.teamsAppClientId = clientIdValue as string;
-    }
 
-    const clientSecretValue: ConfigValue = context.configOfOtherPlugins
+    teamsBotConfig.teamsAppClientSecret = context.configOfOtherPlugins
       .get(PluginAAD.PLUGIN_NAME)
       ?.get(PluginAAD.CLIENT_SECRET);
-    if (clientSecretValue) {
-      this.teamsAppClientSecret = clientSecretValue as string;
-    }
 
-    const tenantIdValue: ConfigValue = context.configOfOtherPlugins
+    teamsBotConfig.teamsAppTenant = context.configOfOtherPlugins
       .get(PluginSolution.PLUGIN_NAME)
       ?.get(PluginSolution.M365_TENANT_ID);
-    if (tenantIdValue) {
-      this.teamsAppTenant = tenantIdValue as string;
-    }
 
-    const applicationIdUrisValue: ConfigValue = context.configOfOtherPlugins
+    teamsBotConfig.applicationIdUris = context.configOfOtherPlugins
       .get(PluginAAD.PLUGIN_NAME)
       ?.get(PluginAAD.APPLICATION_ID_URIS);
-    if (applicationIdUrisValue) {
-      this.applicationIdUris = applicationIdUrisValue as string;
-    }
 
     const capabilities = (context.projectSettings?.solutionSettings as AzureSolutionSettings)
       .capabilities;
 
-    if (capabilities?.includes(PluginActRoles.Bot) && !this.actRoles.includes(PluginActRoles.Bot)) {
-      this.actRoles.push(PluginActRoles.Bot);
+    if (
```

Comment:
why don't use set instead of array so that it can simplify the logic of checking existence?

Reply:
Will keep it for now